### PR TITLE
quic: throw proper errors from c++

### PR DESCRIPTION
**Pull request by jasnell**

### CODE REVIEW

File: src/node_errors.h [hunk 1]
```diff
@@ -102,6 +114,18 @@ void PrintErrorString(const char* format, ...);
...
+# define QUIC_PREDEFINED_ERROR_MESSAGES(V)                                   \
+  V(ERR_QUIC_CANNOT_SET_GROUPS, "Cannot set groups")                         \
+  V(ERR_QUIC_FAILURE_SETTING_SNI_CONTEXT, "Failure setting SNI context")
+#else
+# define QUIC_PREDEFINED_ERROR_MESSAGES(V)
+#endif
+
+#define PREDEFINED_ERROR_MESSAGES(V)                                         \
+  BASE_PREDEFINED_ERROR_MESSAGES(V)                                          \
+  QUIC_PREDEFINED_ERROR_MESSAGES(V)
```
Comment (addaleax) [inline src/node_errors.h]: Tbh, I wouldn’t use conditionals here – everything that happens here happens in inline functions, so if the errors aren’t used, there’s zero overhead.

File: src/quic/node_quic.cc [hunk 2]
```diff
@@ -92,9 +93,10 @@ void QuicInitSecureContext(const FunctionCallbackInfo<Value>& args) {
 
   InitializeSecureContext(sc, side);
 
-  // TODO(@jasnell): Throw a proper node.js error with code
-  if (!crypto::SetGroups(sc, *groups))
-    return env->ThrowError("Failed to set groups");
+  if (!crypto::SetGroups(sc, *groups)) {
+    Isolate* isolate = env->isolate();
+    isolate->ThrowException(ERR_QUIC_CANNOT_SET_GROUPS(isolate));
```
Comment (addaleax) [inline src/quic/node_quic.cc]:
```suggestion
    THROW_ERR_QUIC_CANNOT_SET_GROUPS(env);
```

(afaik that should work)

File: src/quic/node_quic_session.cc [hunk 3]
```diff
@@ -970,8 +972,10 @@ void QuicCryptoContext::OnOCSPDone(
     if (!err) {
       unsigned long err = ERR_get_error();  // NOLINT(runtime/int)
       if (!err) {
-        // TODO(@jasnell): revisit to throw a proper error here
-        return session_->env()->ThrowError("CertCbDone");
+        Isolate* isolate = session_->env()->isolate();
+        isolate->ThrowException(
+            ERR_QUIC_FAILURE_SETTING_SNI_CONTEXT(isolate));
```
Comment (addaleax) [inline src/quic/node_quic_session.cc]: ditto here, you should be able to use the `THROW_...` variant

**Comment by jasnell** [comment]
Landed

**Pull request closed without merging**